tests/unittests/std_interoperability.cpp

```diff
@@ -895,32 +1090,56 @@ TEST_CASE("Collection and std iterator adaptors", "[collection][container][adapt
-    DOCUMENTED_STATIC_FAILURE(std::input_iterator<iterator>);
+    STATIC_REQUIRE(std::input_iterator<iterator>);
 #endif
-    // TODO add more checks here
+    STATIC_REQUIRE(std::is_same_v<iterator::reference, std::move_iterator<iterator>::reference>);
     // const_iterator
     STATIC_REQUIRE(traits::has_iterator_v<CollectionType>);
-    DOCUMENTED_STATIC_FAILURE(traits::has_iterator_category_v<std::iterator_traits<iterator>>);
-    // STATIC_REQUIRE(
-    //    std::is_base_of_v<std::input_iterator_tag, std::iterator_traits<iterator>::iterator_category>);
+    STATIC_REQUIRE(traits::has_iterator_category_v<std::iterator_traits<const_iterator>>);
+    STATIC_REQUIRE(std::is_base_of_v<std::input_iterator_tag, std::iterator_traits<const_iterator>::iterator_category>);
 #if (__cplusplus >= 202002L)
-    DOCUMENTED_STATIC_FAILURE(std::input_iterator<iterator>);
+    STATIC_REQUIRE(std::input_iterator<const_iterator>);
 #endif
-    // TODO add more checks here
+    STATIC_REQUIRE(std::is_same_v<const_iterator::reference, std::move_iterator<const_iterator>::reference>);
   }
 
 #if (__cplusplus >= 202002L)
   SECTION("Counted iterator") {
     // iterator
-    DOCUMENTED_STATIC_FAILURE(std::input_or_output_iterator<iterator>);
-    // TODO add runtime checks
+    STATIC_REQUIRE(std::input_or_output_iterator<iterator>);
+    {
+      auto coll = CollectionType();
+      coll.create().cellID(42);
+      auto counted = std::counted_iterator(coll.begin(), coll.size());
+      REQUIRE(counted != std::default_sentinel);
+      REQUIRE(counted.count() == 1);
+      REQUIRE((*counted).cellID() == 42);
+      DOCUMENTED_STATIC_FAILURE(std::contiguous_iterator<iterator>); // counted-> requires std::contiguous_iterator
+      // REQUIRE(counted->cellID() == 42);
+      DOCUMENTED_STATIC_FAILURE(traits::has_member_of_pointer_v<const iterator>); // can't base()->() because
+                                                                                  // base() returns const object
+      // REQUIRE(counted.base()->cellID() == 42);
+      REQUIRE(++counted == std::default_sentinel);
+    }
     // const_iterator
-    DOCUMENTED_STATIC_FAILURE(std::input_or_output_iterator<const_iterator>);
-    // TODO add runtime checks
+    STATIC_REQUIRE(std::input_or_output_iterator<const_iterator>);
+    {
+      auto coll = CollectionType();
+      coll.create().cellID(42);
+      auto counted = std::counted_iterator(coll.cbegin(), coll.size());
+      REQUIRE(counted != std::default_sentinel);
+      REQUIRE(counted.count() == 1);
+      REQUIRE((*counted).cellID() == 42);
+      DOCUMENTED_STATIC_FAILURE(std::contiguous_iterator<iterator>); // counted-> requires std::contiguous_iterator
+      // REQUIRE(counted->cellID() == 42)
+      DOCUMENTED_STATIC_FAILURE(traits::has_member_of_pointer_v<const const_iterator>); // can't base()->() because
+                                                                                        // base() returns const object
```

Comment:
I don't know enough about the details here, so this is probably a stupid question. Is this something where we are currently missing some syntax, or is this something that is conceptually not possible?

Reply:
That's a very good question :smile: 

It's conceptual. A collection is storing data objects but we want iterators to give user objects. For dereference it's easy becasue a user object can be returned by value, no problem. But `->` operator should return a raw pointer so its pointee must exist somewhere. It can't point to the collection storage because there are no user objects there. So as a workaround the iterators have an additional user object type member to which we can point.
The question is when do we update that member? During `++` would break `end()` (out-of-range access to collection storage to get data object). So this must be during `->` hence it can't be `const`

Strictly speaking there is no requirement for `operator->()` to be `const` in general. It's just the standard adaptors using`const` for their `->` and `base()` for whatever reason. But `*it` and `(*it).member` still works so it's not impossible to use them, just a little confusing